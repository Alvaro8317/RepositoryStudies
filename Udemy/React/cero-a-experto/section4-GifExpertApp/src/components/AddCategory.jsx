import { useState } from 'react';

/*
 * Cada componente puede tener sus propios estados, es decir, sus propios hooks o states
 * Por defecto un form al darle submit o enter, refresca la página, se puede evitar con preventDefault()
 * Se usa aria-label para encontrarlo con react testing library
 */
import PropTypes from 'prop-types';
export const AddCategory = ({ onAddCategories, onNewCategory }) => {
  if (!onAddCategories)
    console.log('Using communications from son to father components');
  const [inputValue, setInputValue] = useState('');
  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };
  const onSubmit = (event) => {
    event.preventDefault();
    if (inputValue.trim().length <= 1) return;
    /* Se realiza la inserción desde el componente hijo y no desde el padre, no se recomienda porque se desconoce la implementació y toca implementar la lógica desde los hijos, si se tiene 10 hijos, toca repetir esta misma lógica 10 veces */
    // onAddCategories((categories) => [...categories, inputValue]);
    /* Se realiza la inserción enviandole al componente padre el valor */
    setInputValue('');
    onNewCategory(inputValue.trim());
  };
  return (
    <form onSubmit={onSubmit} aria-label='form'>
      <input
        type='text'
        placeholder='Search gifs'
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onAddCategories: PropTypes.func,
  onNewCategory: PropTypes.func.isRequired,
};
