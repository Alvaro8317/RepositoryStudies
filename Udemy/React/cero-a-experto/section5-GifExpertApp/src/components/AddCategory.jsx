import { useState } from 'react';

/*
 * Cada componente puede tener sus propios estados, es decir, sus propios hooks o states
 * Por defecto un form al darle submit o enter, refresca la página, se puede evitar con preventDefault()
 */
import PropTypes from 'prop-types';
export const AddCategory = ({ onAddCategories, onNewCategory }) => {
  const [inputValue, setInputValue] = useState('');
  const onInputChange = ({ target }) => {
    setInputValue(target.value);
  };
  const onSubmit = (event) => {
    event.preventDefault();
    if (!onAddCategories)
      console.log('Using communications from son to father components');
    if (inputValue.trim().length <= 1) return;
    /* Se realiza la inserción desde el componente hijo y no desde el padre, no se recomienda porque se desconoce la implementació y toca implementar la lógica desde los hijos, si se tiene 10 hijos, toca repetir esta misma lógica 10 veces */
    // onAddCategories((categories) => [...categories, inputValue]);
    /* Se realiza la inserción enviandole al componente padre el valor */
    onNewCategory(inputValue.trim());
    setInputValue('');
  };
  return (
    <form onSubmit={onSubmit}>
      <input
        type='text'
        placeholder='Search gifs'
        value={inputValue}
        onChange={onInputChange}
      />
    </form>
  );
};

AddCategory.propTypes = {
  onAddCategories: PropTypes.func,
  onNewCategory: PropTypes.func.isRequired,
};
