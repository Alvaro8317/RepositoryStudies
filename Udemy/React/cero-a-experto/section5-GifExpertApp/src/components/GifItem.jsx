import PropTypes from 'prop-types';
export const GifItem = ({ title, url }) => {
  return (
    <div className='card'>
      <img src={url} alt={title} />
      <p> {title} </p>
      {/* <ol>
        <li key={image.id}>{image.title}</li>
      </ol> */}
    </div>
  );
};

GifItem.propTypes = {
  title: PropTypes.string.isRequired,
  url: PropTypes.string.isRequired,
};
