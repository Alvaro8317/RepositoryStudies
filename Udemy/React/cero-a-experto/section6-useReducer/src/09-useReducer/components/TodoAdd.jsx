import PropTypes from 'prop-types';
import { useForm } from '../../hooks/useForm';

export const TodoAdd = ({ onHandleNewTodo }) => {
  const { description, onInputChange, onResetForm } = useForm({
    description: '',
  });
  const onSubmit = (event) => {
    event.preventDefault()
    if (description.length <= 1) return;
    onHandleNewTodo({
      id: new Date().getTime(),
      description: description,
      done: false
    })
    onResetForm()
  }
  return (
    <form onSubmit={onSubmit}>
      <input type="text" placeholder="¿Qué hay que hacer?" name='description' className="form-control" value={description} onChange={onInputChange} />
      <button type="submit" className="btn btn-outline-primary mt-1">Agregar</button>
      <button onClick={onResetForm} className='btn btn-danger mt-1'>Resetear el campo</button>
    </form>
  )
}

TodoAdd.propTypes = {
  onHandleNewTodo: PropTypes.func.isRequired,
}
/* Forma manual */
// const [descriptionToDo, setDescriptionToDo] = useState("")
// const onInputChange = ({ target }) => {
//   setDescriptionToDo(target.value)
// }
// const onSubmit = (event) => {
//   event.preventDefault()
//   onHandleNewTodo({
//     id: new Date().getTime(),
//     description: descriptionToDo,
//     done: false
//   })
// }
// return (
//   <form onSubmit={onSubmit}>
//     <input type="text" placeholder="¿Qué hay que hacer?" className="form-control" value={descriptionToDo} onChange={onInputChange} />
//     <button type="submit" className="btn btn-outline-primary mt-1">Agregar</button>
//   </form>
// )
/* Usando customHook */