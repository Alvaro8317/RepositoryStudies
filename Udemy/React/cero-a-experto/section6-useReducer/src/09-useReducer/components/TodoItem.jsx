import PropTypes from 'prop-types';
export const TodoItem = ({ todo, onHandleTodo, onToggleTodo }) => {
  return (
    <li className="list-group-item d-flex justify-content-between">
      <span className={`align-self-center ${(todo.done) ? 'text-decoration-line-through' : ' '}`} onClick={() => onToggleTodo(todo.id)}>
        {todo.description}
      </span>
      <button className="btn btn-danger" onClick={() => onHandleTodo(todo.id)}>Borrar</button>
    </li>
  )
}

TodoItem.propTypes = {
  todo: PropTypes.object.isRequired,
  onHandleTodo: PropTypes.func.isRequired,
  onToggleTodo: PropTypes.func.isRequired
}