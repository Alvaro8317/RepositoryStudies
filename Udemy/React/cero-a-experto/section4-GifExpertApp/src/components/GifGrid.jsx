/*
 * Dejar una función port fuera del componente es mejor para evitar que se renderice la función
 * useEffect es para lanzar efectos secundarios, el primer parámetro es una función que se quiere ejecutar y el segundo, que es opcional - obligatorio que son las dependencias
 * En JSX no se usa class porque se puede confundir, para clases CSS se deja className
 * Esparcir las propiedades
 */

import PropTypes from 'prop-types';
import { useFetchGifs } from '../hooks/useFetchGifs';
// import { GifItem } from './GifItem'; Sin importación barril
import { GifItem } from './GifItem';
/*
 * Los custom hooks se recomienda que empiecen por use
 */

export const GifGrid = ({ category }) => {
  const { images, isLoading } = useFetchGifs(category);
  console.log(images, isLoading);

  return (
    <>
      <h3>{category}</h3>
      {/* Tecnica para dejar un cargando: */}
      {isLoading && <h2>Cargando...</h2>}
      <div className='card-grid'>
        {images.map((image) => (
          <GifItem key={image.id} {...image} />
        ))}
      </div>
    </>
  );
};

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
