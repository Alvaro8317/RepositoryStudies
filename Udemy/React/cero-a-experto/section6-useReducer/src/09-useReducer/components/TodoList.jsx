import { TodoItem } from './TodoItem'
import PropTypes from 'prop-types';
export const TodoList = ({ todos = [], onHandleTodo, onToggleTodo: handleToggleTodo }) => {
  return (
    <ul className="list-group">
      {
        todos.map(todoElement =>
        (
          <TodoItem
            key={todoElement.id}
            todo={todoElement}
            onHandleTodo={onHandleTodo}
            onToggleTodo={handleToggleTodo}
          />)
        )
      }
    </ul>
  )
}

TodoList.propTypes = {
  todos: PropTypes.array.isRequired,
  onHandleTodo: PropTypes.func.isRequired,
  onToggleTodo: PropTypes.func.isRequired
}