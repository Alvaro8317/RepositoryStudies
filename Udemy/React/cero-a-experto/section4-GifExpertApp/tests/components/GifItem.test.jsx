import { render, screen } from '@testing-library/react';
import { GifItem } from '../../src/components/GifItem';

describe('Tests of GifItem Component', () => {
  const title = 'metal gear art GIF';
  const url = 'https://giphy.com/gifs/oktotally-art-pixel-jQQ854R8tx3wI';
  test('should match the snapshot', () => {
    const { container } = render(<GifItem title={title} url={url} />);
    expect(container).toMatchSnapshot();
  });
  test('should show the image with the url and the title', () => {
    render(<GifItem title={title} url={url} />);
    // screen.debug();
    expect(screen.getByRole('img').src).toBe(url);
    expect(screen.getByRole('img').alt).toBe(title);
  });
  test('should show the title in the component', () => {
    render(<GifItem title={title} url={url} />);
    expect(screen.getByText(title)).toBeTruthy();
  });
});
