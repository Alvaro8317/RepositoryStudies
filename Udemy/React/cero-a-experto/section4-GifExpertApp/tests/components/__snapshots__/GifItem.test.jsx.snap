// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests of GifItem Component should match the snapshot 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt="metal gear art GIF"
      src="https://giphy.com/gifs/oktotally-art-pixel-jQQ854R8tx3wI"
    />
    <p>
       
      metal gear art GIF
       
    </p>
  </div>
</div>
`;
